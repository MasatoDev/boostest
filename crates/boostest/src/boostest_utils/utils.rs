use anyhow::{anyhow, Result};
use std::path::{Path, PathBuf};

use glob::glob;
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::{self, File};
use std::io::{self, prelude::*};

pub fn read(path: &Path) -> io::Result<String> {
    let mut f = File::open(path)?;
    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

pub fn read_matching_files(
    patterns: Vec<String>,
    out_file_name: &String,
) -> anyhow::Result<HashMap<PathBuf, String>> {
    let mut contents: HashMap<PathBuf, String> = HashMap::new();

    for pattern in &patterns {
        let result_glob = glob(pattern)?;

        for entry in result_glob {
            let path = entry?;

            if path
                .file_name()
                .unwrap_or(OsStr::new("none_file_name"))
                .to_string_lossy()
                .contains(out_file_name.as_str())
            {
                continue;
            }

            if path.is_file() {
                contents.insert(path.clone(), fs::read_to_string(path)?);
            }
        }
    }

    Ok(contents)
}

pub fn find_boostest_json_recursive(dir: PathBuf) -> anyhow::Result<PathBuf> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(file_name) = path.file_name() {
                if file_name == "boostest.setting.json" {
                    return Ok(path);
                }
            }
        } else if path.is_dir() {
            if let Ok(found_path) = find_boostest_json_recursive(path) {
                return Ok(found_path);
            }
        }
    }

    Err(anyhow!("boostest.json not found"))
}

/**
 * Normalize and resolve the path.
 * parent: "hoge/a/b/c", relative: "../../d/e/f" => "hoge/a/d/e/f"
 */
pub fn normalize_and_resolve_path(parent: &Path, relative: &Path) -> Result<PathBuf, String> {
    let combined_path = parent.join(relative);
    let resolved_path = combined_path
        .components()
        .fold(PathBuf::new(), |mut acc, comp| {
            match comp {
                std::path::Component::ParentDir => {
                    acc.pop();
                }
                std::path::Component::CurDir => {}
                other => acc.push(other),
            }
            acc
        });

    if resolved_path.is_absolute() {
        Ok(resolved_path)
    } else {
        Err("The resolved path is not absolute.".to_string())
    }
}
