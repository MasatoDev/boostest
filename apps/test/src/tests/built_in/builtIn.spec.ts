import { runSnapshotTest } from "../utils";

/**
 *
 *
 * Union type
 *
 *
 */
import {
  TSAliasMixUnionObjType,
  TSAliasMixUnionType,
  TSAliasStringUnionType,
  TSInterfaceMixUnionType,
  TSInterfaceStringUnionType,
} from "../../types/ts_types/union";
// import { boostestTSAliasMixUnionType } from "./boostest_output/boostestTSAliasMixUnionType";
// import { boostestTSAliasStringUnionType } from "./boostest_output/boostestTSAliasStringUnionType";
// import { boostestTSInterfaceMixUnionType } from "./boostest_output/boostestTSInterfaceMixUnionType";
// import { boostestTSInterfaceStringUnionType } from "./boostest_output/boostestTSInterfaceStringUnionType";
// import { boostestTSAliasMixUnionObjType } from "./boostest_output/boostestTSAliasMixUnionObjType";
import {
  BuiltInClass,
  BuiltInInterface,
  BuiltInType,
} from "@/ts_types/built_in_class";
import { boostestTSAliasBuiltInType } from "./boostest_output/boostestTSAliasBuiltInType";
import { boostestInterfaceBuiltInType } from "./boostest_output/boostestInterfaceBuiltInType";
// import { boostestClassBuiltInType } from "./boostest_output/boostestClassBuiltInType";

describe("BuiltIn Tests", () => {
  runSnapshotTest(
    "type alias BuiltIn",
    boostestTSAliasBuiltInType<BuiltInType>(),
  );
  runSnapshotTest(
    "interface BuiltIn",
    boostestInterfaceBuiltInType<BuiltInInterface>(),
  );
  // TODO:
  // runSnapshotTest("class BuiltIn", boostestClassBuiltInType<BuiltInClass>());
});
