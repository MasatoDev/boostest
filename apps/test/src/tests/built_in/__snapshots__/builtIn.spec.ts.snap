// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuiltIn Tests type alias BuiltIn matches snapshot 1`] = `"{"date":"test string data","set":{},"map":{},"array":[],"object":{"constructor":"() => { }","toString":"() => {\\n                return \\"test string data\\";\\n            }","toLocaleString":"() => {\\n                return \\"test string data\\";\\n            }","valueOf":"() => {\\n                return {};\\n            }","hasOwnProperty":"(v) => {\\n                return true;\\n            }","isPrototypeOf":"(v) => {\\n                return true;\\n            }","propertyIsEnumerable":"(v) => {\\n                return true;\\n            }"},"string":{"0":"test string data","toString":"() => {\\n                return \\"test string data\\";\\n            }","charAt":"(pos) => {\\n                return \\"test string data\\";\\n            }","charCodeAt":"(index) => {\\n                return 10;\\n            }","concat":"(...strings) => {\\n                return \\"test string data\\";\\n            }","indexOf":"(searchString, position) => {\\n                return 10;\\n            }","lastIndexOf":"(searchString, position) => {\\n                return 10;\\n            }","localeCompare":"(that) => {\\n                return 10;\\n            }","match":"(regexp) => {\\n                return \\"any\\";\\n            }","replace":"(searchValue, replaceValue) => {\\n                return \\"test string data\\";\\n            }","search":"(regexp) => {\\n                return 10;\\n            }","slice":"(start, end) => {\\n                return \\"test string data\\";\\n            }","split":"(separator, limit) => {\\n                return [];\\n            }","substring":"(start, end) => {\\n                return \\"test string data\\";\\n            }","toLowerCase":"() => {\\n                return \\"test string data\\";\\n            }","toLocaleLowerCase":"(locales) => {\\n                return \\"test string data\\";\\n            }","toUpperCase":"() => {\\n                return \\"test string data\\";\\n            }","toLocaleUpperCase":"(locales) => {\\n                return \\"test string data\\";\\n            }","trim":"() => {\\n                return \\"test string data\\";\\n            }","length":10,"substr":"(from, length) => {\\n                return \\"test string data\\";\\n            }","valueOf":"() => {\\n                return \\"test string data\\";\\n            }"},"number":{"toString":"(radix) => {\\n                return \\"test string data\\";\\n            }","toFixed":"(fractionDigits) => {\\n                return \\"test string data\\";\\n            }","toExponential":"(fractionDigits) => {\\n                return \\"test string data\\";\\n            }","toPrecision":"(precision) => {\\n                return \\"test string data\\";\\n            }","valueOf":"() => {\\n                return 10;\\n            }","toLocaleString":"(locales, options) => {\\n                return \\"test string data\\";\\n            }"},"boolean":{"valueOf":"() => {\\n                return true;\\n            }"},"symbol":{"toString":"() => {\\n                return \\"test string data\\";\\n            }","valueOf":"() => {\\n                return Symbol();\\n            }"},"function":"() => { }"}"`;
